add `--component-type` and `--string-encoding` options

dicej opened the pull request:
Per https://github.com/bytecodealliance/wit-bindgen/issues/994, this allows all or part of the component type to be specified as one or more WIT files rather than binary-encoded custom sections.  This is akin to inserting a `wasm-tools component embed` command between the linking and componentization stages.

alexcrichton commented on src/lib.rs:
```diff
@@ -542,6 +571,40 @@ impl App {
+        if !self.component.component_types.is_empty() {
+            let mut merged = None::<(Resolve, WorldId)>;
+            for wit_file in &self.component.component_types {
+                let mut resolve = Resolve::default();
```
Could this be lifted out of the loop to have everything pushed into the same `Resolve`? That would remove the need to use `merge`, only `merge_worlds` would be needed

alexcrichton replied:
I talked with Joel a bit more about this and we concluded that the way the code is currently structured is the way to go.

alexcrichton commented on src/lib.rs:
```diff
@@ -542,6 +571,40 @@ impl App {
+                resolve
+                    .push_path(wit_file)
+                    .with_context(|| format!("unable to add component type {wit_file:?}"))?;
+                if resolve.worlds.len() != 1 {
+                    bail!(
+                        "expected exactly 1 world in {wit_file:?}; found {}",
+                        resolve.worlds.len()
+                    );
+                }
```
I'm hesitant about this as I'm afraid that worlds will sneak in pretty quickly. Could this perhaps stick to the "path plus `--world`" combo of other tools? That way this could use `Resolve::select_world`. It would preclude multiple `--component-type` arguments, however, so that would only work if that's ok for your use case.

dicej replied:
That would be difficult for two reasons:
- The goal is to be able to check this WIT file into source control and pass it to `wasm-component-ld` as easily as we were previously passing `_component_type.o` files, with no further information needed.  Not sure where the world name would be stored.
- There will definitely be more than one of these files in general.  At least one will be bundled as part of the .NET runtime, and another coming from any bindings generated by the application developer (plus any which are part of SDKs that application is using).

dicej replied:
Perhaps we should reconsider using WAT instead of WIT for this?  With WAT there isn't any notion of worlds -- just top-level imports and exports, so all the questions about which world and which string encoding go away.

dicej replied:
Or else: stick with WIT but ensure that `wit-bindgen` only ever emits exactly one world -- the one it was provided as input (i.e. flatten any `include` statements that refer to other worlds).

alexcrichton replied:
Joel and I talked more about this and the conclusion was:

* The general structure here will stay except that `select_world` will be used to provide some more future flexibility.
* The generated `*.wit` file in the bindgen generators will be generated differently to work with this.

alexcrichton reviewed (commented):
Thanks! Could you also be sure to add a test case too?

dicej commented:
I've updated the code here to use `Resolve::select_world` and updated `wit-bindgen` to output only single-world WIT files.  Working on a test case now.

alexcrichton commented on src/lib.rs:
```diff
@@ -542,6 +571,43 @@ impl App {
+                    .with_context(|| format!("unable to add component type {wit_file:?}"))?;
+
+                let world = resolve.select_world(
+                    &resolve
+                        .packages
+                        .iter()
+                        .map(|(id, _)| id)
+                        .collect::<Vec<_>>(),
```
Oh this argument should be the return value of `push_path`, or the first element of its tuple

dicej replied:
Done; thanks!

alexcrichton reviewed (approved):
Looks great! One minor comment but otherwise good to go
